RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -11,40 +10,65 @@
+        return Flux.using(
+            () -> openIndexReader(indexDirectoryPath),
+            reader -> {
+                logger.info(reader.maxDoc() + " documents found in the current Lucene index");
 
-        try (FSDirectory directory = FSDirectory.open(indexDirectoryPath);
-             IndexReader reader = DirectoryReader.open(directory)) {
-
-            // Add all documents to our output that have the _source field set and filled in
-            for (int i = 0; i < reader.maxDoc(); i++) {
-                Document document = reader.document(i);
-                BytesRef source_bytes = document.getBinaryValue("_source");
-                String id;
-                // TODO Improve handling of missing document id (https://opensearch.atlassian.net/browse/MIGRATIONS-1649)
+                return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
+                .handle((i, sink) -> {
+                    Document doc = getDocument(reader, i);
+                    if (doc != null) { // Skip malformed docs
+                        sink.next(doc);
+                    }
+                }).cast(Document.class);
+            },
+            reader -> { // Close the IndexReader when done
                 try {
-                    id = Uid.decodeId(reader.document(i).getBinaryValue("_id").bytes);
-                } catch (Exception e) {
-                    logger.warn("Unable to parse Document id from Document");
-                    id = "unknown-id";
-                }
-                if (source_bytes == null || source_bytes.bytes.length == 0) { // Skip deleted documents
-                    logger.info("Document " + id + " is deleted or doesn't have the _source field enabled");
-                    continue;
+                    reader.close();
+                } catch (IOException e) {
+                    logger.error("Failed to close IndexReader", e);
                 }
+            }
+        );
+    }
 
-                documents.add(document);
-                logger.info("Document " + id + " read successfully");
+    protected IndexReader openIndexReader(Path indexDirectoryPath) throws IOException {
+        return DirectoryReader.open(FSDirectory.open(indexDirectoryPath));
+    }
+
+    protected Document getDocument(IndexReader reader, int docId) {
+        try {
+            Document document = reader.document(docId);
+            BytesRef source_bytes = document.getBinaryValue("_source");
+            String id;
+            try {
+                id = Uid.decodeId(document.getBinaryValue("_id").bytes);
+            } catch (Exception e) {
+                StringBuilder errorMessage = new StringBuilder();
+                errorMessage.append("Unable to parse Document id from Document.  The Document's Fields: ");
+                document.getFields().forEach(f -> errorMessage.append(f.name()).append(", "));
+                logger.error(errorMessage.toString());
```

Comment:
`logger.atError().setCause(e).setMessage(()->...).log()` will do two more things for you.  1) get the exception and its backtrace into the logs and 2) use the fluent style, where everything within '...' will _only_ be evaluated when you're logging that level.  It can make your log statements tighter (all one statement rather than 4 as they are here) and much more efficient since work can often be elided.  Even if you stay at warn/error, I'd like to routinely filter the repo for usages of immediate logging because its performance hit can be the single greatest impact on a program.

Reply:
Consider PII for ERROR.  I think that it's fair, but you should call it out... maybe PII possible loggers should have their own logger name convention so that operators could easily mask them out if necessary.